File: bot.py
import os
import logging
import sqlite3
from datetime import datetime
from typing import Dict, Any

from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import (
    Application, CommandHandler, CallbackQueryHandler, 
    MessageHandler, filters, ContextTypes, ConversationHandler
)
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Configure logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.INFO
)
logger = logging.getLogger(__name__)

# Conversation states
FAMILY, CATEGORY, AMOUNT, PERIOD, CONTACT = range(5)

# Predefined families
FAMILIES = [
    "Famiglia Rossi",
    "Famiglia Bianchi", 
    "Famiglia Verdi",
    "Famiglia Neri",
    "Famiglia Blu"
]

# Categories
CATEGORIES = [
    "Quota Mensile",
    "Quota Iscrizione", 
    "Altro"
]

class TransactionBot:
    def __init__(self):
        self.bot_token = os.getenv('TELEGRAM_BOT_TOKEN')
        self.allowed_group_id = os.getenv('ALLOWED_GROUP_ID')
        self.db_path = os.getenv('DATABASE_PATH', 'transactions.db')
        self.init_database()
    
    def init_database(self):
        """Initialize SQLite database for storing transactions"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS transactions (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                family TEXT NOT NULL,
                category TEXT NOT NULL,
                amount REAL NOT NULL,
                period TEXT NOT NULL,
                contact TEXT NOT NULL,
                user_id INTEGER NOT NULL,
                username TEXT,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        ''')
        
        conn.commit()
        conn.close()
    
    def is_authorized(self, update: Update) -> bool:
        """Check if the user is authorized to use the bot"""
        if not self.allowed_group_id:
            return True  # If no group restriction, allow all
        
        chat_id = str(update.effective_chat.id)
        return chat_id == self.allowed_group_id
    
    async def start(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
        """Start the conversation and ask for family selection"""
        if not self.is_authorized(update):
            await update.message.reply_text("❌ Questo bot può essere utilizzato solo nel gruppo autorizzato.")
            return ConversationHandler.END
        
        keyboard = []
        for family in FAMILIES:
            keyboard.append([InlineKeyboardButton(family, callback_data=f"family_{family}")])
        
        reply_markup = InlineKeyboardMarkup(keyboard)
        
        await update.message.reply_text(
            "💰 **Registrazione Transazione**\n\n"
            "Seleziona la famiglia:",
            reply_markup=reply_markup,
            parse_mode='Markdown'
        )
        
        return FAMILY
    
    async def select_family(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
        """Handle family selection"""
        query = update.callback_query
        await query.answer()
        
        family = query.data.replace("family_", "")
        context.user_data['family'] = family
        
        # Show category selection
        keyboard = []
        for category in CATEGORIES:
            keyboard.append([InlineKeyboardButton(category, callback_data=f"category_{category}")])
        
        reply_markup = InlineKeyboardMarkup(keyboard)
        
        await query.edit_message_text(
            f"👨‍👩‍👧‍👦 **Famiglia selezionata:** {family}\n\n"
            "Seleziona la categoria:",
            reply_markup=reply_markup,
            parse_mode='Markdown'
        )
        
        return CATEGORY
    
    async def select_category(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
        """Handle category selection"""
        query = update.callback_query
        await query.answer()
        
        category = query.data.replace("category_", "")
        context.user_data['category'] = category
        
        await query.edit_message_text(
            f"👨‍👩‍👧‍👦 **Famiglia:** {context.user_data['family']}\n"
            f"📂 **Categoria:** {category}\n\n"
            "💰 Inserisci l'importo in Euro (es. 25.50):"
        )
        
        return AMOUNT
    
    async def get_amount(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
        """Handle amount input"""
        try:
            amount = float(update.message.text.replace(',', '.'))
            if amount <= 0:
                raise ValueError("L'importo deve essere positivo")
            
            context.user_data['amount'] = amount
            
            await update.message.reply_text(
                f"👨‍👩‍👧‍👦 **Famiglia:** {context.user_data['family']}\n"
                f"📂 **Categoria:** {context.user_data['category']}\n"
                f"💰 **Importo:** €{amount:.2f}\n\n"
                "📅 Inserisci il periodo (formato: YYYY-MM-DD, es. 2024-01-15):"
            )
            
            return PERIOD
            
        except ValueError:
            await update.message.reply_text(
                "❌ Formato non valido. Inserisci un importo numerico valido (es. 25.50):"
            )
            return AMOUNT
    
    async def get_period(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
        """Handle period input"""
        try:
            period = update.message.text.strip()
            # Validate date format
            datetime.strptime(period, '%Y-%m-%d')
            
            context.user_data['period'] = period
            
            await update.message.reply_text(
                f"👨‍👩‍👧‍👦 **Famiglia:** {context.user_data['family']}\n"
                f"📂 **Categoria:** {context.user_data['category']}\n"
                f"💰 **Importo:** €{context.user_data['amount']:.2f}\n"
                f"📅 **Periodo:** {period}\n\n"
                "👤 Inserisci il username del contatto (es. @username):"
            )
            
            return CONTACT
            
        except ValueError:
            await update.message.reply_text(
                "❌ Formato data non valido. Usa il formato YYYY-MM-DD (es. 2024-01-15):"
            )
            return PERIOD
    
    async def get_contact(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
        """Handle contact input and save transaction"""
        contact = update.message.text.strip()
        
        # Validate username format
        if not contact.startswith('@'):
            contact = '@' + contact
        
        # Save transaction to database
        self.save_transaction(update, context, contact)
        
        # Send notification to the specified contact
        await self.send_notification(update, context, contact)
        
        # Confirm to user
        await update.message.reply_text(
            f"✅ **Transazione registrata con successo!**\n\n"
            f"👨‍👩‍👧‍👦 **Famiglia:** {context.user_data['family']}\n"
            f"📂 **Categoria:** {context.user_data['category']}\n"
            f"💰 **Importo:** €{context.user_data['amount']:.2f}\n"
            f"📅 **Periodo:** {context.user_data['period']}\n"
            f"👤 **Contatto:** {contact}\n\n"
            f"📨 Notifica inviata a {contact}",
            parse_mode='Markdown'
        )
        
        # Clear user data
        context.user_data.clear()
        
        return ConversationHandler.END
    
    def save_transaction(self, update: Update, context: ContextTypes.DEFAULT_TYPE, contact: str):
        """Save transaction to database"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute('''
            INSERT INTO transactions 
            (family, category, amount, period, contact, user_id, username)
            VALUES (?, ?, ?, ?, ?, ?, ?)
        ''', (
            context.user_data['family'],
            context.user_data['category'],
            context.user_data['amount'],
            context.user_data['period'],
            contact,
            update.effective_user.id,
            update.effective_user.username or 'N/A'
        ))
        
        conn.commit()
        conn.close()
    
    async def send_notification(self, update: Update, context: ContextTypes.DEFAULT_TYPE, contact: str):
        """Send notification to the specified contact"""
        try:
            notification_message = (
                f"🔔 **Nuova Transazione Registrata**\n\n"
                f"👨‍👩‍👧‍👦 **Famiglia:** {context.user_data['family']}\n"
                f"📂 **Categoria:** {context.user_data['category']}\n"
                f"💰 **Importo:** €{context.user_data['amount']:.2f}\n"
                f"📅 **Periodo:** {context.user_data['period']}\n"
                f"👤 **Registrato da:** @{update.effective_user.username or 'utente'}\n"
                f"⏰ **Data:** {datetime.now().strftime('%Y-%m-%d %H:%M')}"
            )
            
            # Try to send to the contact
            await context.bot.send_message(
                chat_id=contact,
                text=notification_message,
                parse_mode='Markdown'
            )
            
        except Exception as e:
            logger.error(f"Failed to send notification to {contact}: {e}")
            # Notify the user that notification failed
            await update.message.reply_text(
                f"⚠️ Transazione salvata ma non è stato possibile inviare la notifica a {contact}.\n"
                f"Verifica che l'username sia corretto."
            )
    
    async def cancel(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
        """Cancel the conversation"""
        await update.message.reply_text("❌ Operazione annullata.")
        context.user_data.clear()
        return ConversationHandler.END
    
    async def help_command(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        """Send help message"""
        if not self.is_authorized(update):
            await update.message.reply_text("❌ Questo bot può essere utilizzato solo nel gruppo autorizzato.")
            return
        
        help_text = (
            "🤖 **Bot di Gestione Transazioni**\n\n"
            "**Comandi disponibili:**\n"
            "/start - Inizia una nuova registrazione transazione\n"
            "/help - Mostra questo messaggio di aiuto\n\n"
            "**Come usare:**\n"
            "1. Usa /start per iniziare\n"
            "2. Segui le istruzioni per inserire i dati\n"
            "3. La transazione verrà salvata e una notifica inviata al contatto specificato"
        )
        
        await update.message.reply_text(help_text, parse_mode='Markdown')
    
    def run(self):
        """Run the bot"""
        if not self.bot_token:
            logger.error("TELEGRAM_BOT_TOKEN not found in environment variables")
            return
        
        # Create application
        application = Application.builder().token(self.bot_token).build()
        
        # Create conversation handler
        conv_handler = ConversationHandler(
            entry_points=[CommandHandler('start', self.start)],
            states={
                FAMILY: [CallbackQueryHandler(self.select_family, pattern='^family_')],
                CATEGORY: [CallbackQueryHandler(self.select_category, pattern='^category_')],
                AMOUNT: [MessageHandler(filters.TEXT & ~filters.COMMAND, self.get_amount)],
                PERIOD: [MessageHandler(filters.TEXT & ~filters.COMMAND, self.get_period)],
                CONTACT: [MessageHandler(filters.TEXT & ~filters.COMMAND, self.get_contact)],
            },
            fallbacks=[CommandHandler('cancel', self.cancel)],
        )
        
        # Add handlers
        application.add_handler(conv_handler)
        application.add_handler(CommandHandler('help', self.help_command))
        
        # Start the bot
        logger.info("Starting bot...")
        application.run_polling()

if __name__ == '__main__':
    bot = TransactionBot()
    bot.run()

File: src/bot.ts
import TelegramBot from 'node-telegram-bot-api';
import { Database } from './database';
import { UserSession, Transaction, FAMILY_OPTIONS, CATEGORY_OPTIONS } from './types';

export class EnBot {
  private bot: TelegramBot;
  private db: Database;
  private userSessions: Map<number, UserSession> = new Map();
  private allowedGroupId: string;
  private adminUserIds: number[];

  constructor(token: string, allowedGroupId: string, adminUserIds: number[] = []) {
    this.bot = new TelegramBot(token, { polling: true });
    this.db = new Database();
    this.allowedGroupId = allowedGroupId;
    this.adminUserIds = adminUserIds;
    this.setupHandlers();
  }

  private setupHandlers(): void {
    // Start command
    this.bot.onText(/\/start/, (msg) => {
      if (!this.isAllowedChat(msg.chat.id, msg.from?.id)) {
        this.bot.sendMessage(msg.chat.id, '❌ Questo bot può essere utilizzato solo nel gruppo autorizzato o da utenti admin.');
        return;
      }
      this.startTransaction(msg);
    });

    // Help command
    this.bot.onText(/\/help/, (msg) => {
      if (!this.isAllowedChat(msg.chat.id, msg.from?.id)) {
        return;
      }
      this.showHelp(msg);
    });

    // Cancel command
    this.bot.onText(/\/cancel/, (msg) => {
      if (!this.isAllowedChat(msg.chat.id, msg.from?.id)) {
        return;
      }
      this.cancelTransaction(msg);
    });

    // Handle callback queries (inline keyboard buttons)
    this.bot.on('callback_query', (callbackQuery) => {
      if (!callbackQuery.message?.chat.id || !this.isAllowedChat(callbackQuery.message.chat.id, callbackQuery.from.id)) {
        return;
      }
      this.handleCallbackQuery(callbackQuery);
    });

    // Handle text messages
    this.bot.on('message', (msg) => {
      if (!this.isAllowedChat(msg.chat.id, msg.from?.id) || msg.text?.startsWith('/')) {
        return;
      }
      this.handleTextMessage(msg);
    });
  }

  private isAllowedChat(chatId: number, userId?: number): boolean {
    // Allow admin users from any chat
    if (userId && this.adminUserIds.includes(userId)) {
      return true;
    }
    // Allow anyone from the specified group
    return chatId.toString() === this.allowedGroupId;
  }

  private startTransaction(msg: TelegramBot.Message): void {
    const chatId = msg.chat.id;
    const userId = msg.from?.id;

    console.log(`🚀 Starting transaction for user ${userId} in chat ${chatId}`);

    if (!userId) {
      console.log('❌ No userId found, cannot start transaction');
      return;
    }

    // Initialize user session
    this.userSessions.set(userId, {
      chatId,
      userId,
      step: 'family',
      transactionData: {}
    });

    console.log(`✅ Session created for user ${userId}, step: family`);
    this.sendFamilySelection(chatId);
  }

  private sendFamilySelection(chatId: number): void {
    const keyboard = {
      reply_markup: {
        inline_keyboard: FAMILY_OPTIONS.map(family => [
          { text: family, callback_data: `family_${family}` }
        ])
      }
    };

    this.bot.sendMessage(chatId, '👨‍👩‍👧‍👦 Seleziona la famiglia:', keyboard);
  }

  private sendCategorySelection(chatId: number): void {
    const keyboard = {
      reply_markup: {
        inline_keyboard: CATEGORY_OPTIONS.map(category => [
          { text: category, callback_data: `category_${category}` }
        ])
      }
    };

    this.bot.sendMessage(chatId, '📋 Seleziona la categoria:', keyboard);
  }

  private handleCallbackQuery(callbackQuery: TelegramBot.CallbackQuery): void {
    const data = callbackQuery.data;
    const chatId = callbackQuery.message?.chat.id;
    const userId = callbackQuery.from.id;

    if (!chatId || !data) return;

    const session = this.userSessions.get(userId);
    if (!session) return;

    if (data.startsWith('family_')) {
      const family = data.replace('family_', '');
      session.transactionData.family = family;
      session.step = 'category';
      
      this.bot.answerCallbackQuery(callbackQuery.id, { text: `Famiglia selezionata: ${family}` });
      this.sendCategorySelection(chatId);
    } else if (data.startsWith('category_')) {
      const category = data.replace('category_', '');
      session.transactionData.category = category;
      session.step = 'amount';
      
      this.bot.answerCallbackQuery(callbackQuery.id, { text: `Categoria selezionata: ${category}` });
      this.bot.sendMessage(chatId, '💰 Inserisci l\'importo in EUR (es. 25.50):');
    }
  }

  private handleTextMessage(msg: TelegramBot.Message): void {
    const chatId = msg.chat.id;
    const userId = msg.from?.id;
    const text = msg.text;

    console.log(`📨 Received text message from user ${userId}: "${text}"`);

    if (!userId || !text) {
      console.log('❌ Missing userId or text, ignoring message');
      return;
    }

    const session = this.userSessions.get(userId);
    if (!session) {
      console.log(`❌ No session found for user ${userId}`);
      return;
    }

    console.log(`🔄 Processing message for user ${userId}, current step: ${session.step}`);

    switch (session.step) {
      case 'amount':
        console.log(`💰 Handling amount input: "${text}"`);
        this.handleAmountInput(chatId, userId, text, session);
        break;
      case 'period':
        console.log(`📅 Handling period input: "${text}"`);
        this.handlePeriodInput(chatId, userId, text, session);
        break;
      case 'contact':
        console.log(`👤 Handling contact input: "${text}"`);
        this.handleContactInput(chatId, userId, text, session);
        break;
      default:
        console.log(`❓ Unknown step: ${session.step}`);
    }
  }

  private handleAmountInput(chatId: number, userId: number, text: string, session: UserSession): void {
    console.log(`💰 Processing amount input: "${text}" for user ${userId}`);
    const amount = parseFloat(text);
    console.log(`💰 Parsed amount: ${amount}`);
    
    if (isNaN(amount) || amount <= 0) {
      console.log(`❌ Invalid amount: ${amount}`);
      this.bot.sendMessage(chatId, '❌ Inserisci un importo valido in EUR (es. 25.50):');
      return;
    }

    session.transactionData.amount = amount;
    session.step = 'period';
    console.log(`✅ Amount saved: ${amount}, moving to period step`);
    
    this.bot.sendMessage(chatId, '📅 Inserisci il periodo (formato: YYYY-MM-DD, es. 2024-01-15):');
  }

  private handlePeriodInput(chatId: number, userId: number, text: string, session: UserSession): void {
    console.log(`📅 Processing period input: "${text}" for user ${userId}`);
    // Simple date validation
    const dateRegex = /^\d{4}-\d{2}-\d{2}$/;
    if (!dateRegex.test(text)) {
      console.log(`❌ Invalid date format: "${text}"`);
      this.bot.sendMessage(chatId, '❌ Inserisci una data valida nel formato YYYY-MM-DD (es. 2024-01-15):');
      return;
    }

    session.transactionData.period = text;
    session.step = 'contact';
    console.log(`✅ Period saved: ${text}, moving to contact step`);
    
    this.bot.sendMessage(chatId, '👤 Inserisci il username del contatto (es. @username):');
  }

  private handleContactInput(chatId: number, userId: number, text: string, session: UserSession): void {
    console.log(`👤 Processing contact input: "${text}" for user ${userId}`);
    // Validate username format
    if (!text.startsWith('@')) {
      console.log(`❌ Invalid username format: "${text}"`);
      this.bot.sendMessage(chatId, '❌ Inserisci un username valido che inizi con @ (es. @username):');
      return;
    }

    session.transactionData.contact = text;
    console.log(`✅ Contact saved: ${text}, completing transaction`);
    
    // Complete the transaction
    this.completeTransaction(chatId, userId, session);
  }

  private async completeTransaction(chatId: number, userId: number, session: UserSession): Promise<void> {
    try {
      const transaction: Omit<Transaction, 'id'> = {
        family: session.transactionData.family!,
        category: session.transactionData.category!,
        amount: session.transactionData.amount!,
        period: session.transactionData.period!,
        contact: session.transactionData.contact!,
        recordedBy: `@${session.userId}`,
        recordedAt: new Date().toISOString(),
        chatId
      };

      const transactionId = await this.db.saveTransaction(transaction);
      
      // Send confirmation to the group
      const confirmationMessage = `
✅ **Transazione registrata con successo!**

📋 **Dettagli:**
• **Famiglia:** ${transaction.family}
• **Categoria:** ${transaction.category}
• **Importo:** €${transaction.amount}
• **Periodo:** ${transaction.period}
• **Contatto:** ${transaction.contact}
• **Registrato da:** ${transaction.recordedBy}
• **ID Transazione:** #${transactionId}

📤 **Notifica inviata a:** ${transaction.contact}
      `;

      this.bot.sendMessage(chatId, confirmationMessage, { parse_mode: 'Markdown' });

      // Send notification to the contact
      const notificationMessage = `
🔔 **Nuova transazione registrata**

📋 **Dettagli:**
• **Famiglia:** ${transaction.family}
• **Categoria:** ${transaction.category}
• **Importo:** €${transaction.amount}
• **Periodo:** ${transaction.period}
• **Registrato da:** ${transaction.recordedBy}
• **ID Transazione:** #${transactionId}
      `;

      try {
        await this.bot.sendMessage(transaction.contact, notificationMessage, { parse_mode: 'Markdown' });
      } catch (error) {
        console.error('Error sending notification:', error);
        this.bot.sendMessage(chatId, `⚠️ Impossibile inviare la notifica a ${transaction.contact}. Verifica che l'username sia corretto.`);
      }

      // Clear session
      this.userSessions.delete(userId);

    } catch (error) {
      console.error('Error saving transaction:', error);
      this.bot.sendMessage(chatId, '❌ Errore durante il salvataggio della transazione. Riprova.');
    }
  }

  private showHelp(msg: TelegramBot.Message): void {
    const helpMessage = `
🤖 **EnBot - Gestione Transazioni**

**Comandi disponibili:**
• /start - Inizia una nuova transazione
• /help - Mostra questo messaggio di aiuto
• /cancel - Annulla la transazione in corso

**Come utilizzare:**
1. Usa /start per iniziare
2. Segui le istruzioni per inserire:
   • Famiglia (selezione da menu)
   • Categoria (selezione da menu)
   • Importo in EUR
   • Periodo (formato: YYYY-MM-DD)
   • Username del contatto (@username)

**Categorie disponibili:**
• quota mensile
• quota iscrizione
• altro
    `;

    this.bot.sendMessage(msg.chat.id, helpMessage, { parse_mode: 'Markdown' });
  }

  private cancelTransaction(msg: TelegramBot.Message): void {
    const userId = msg.from?.id;
    if (!userId) return;

    const session = this.userSessions.get(userId);
    if (session) {
      this.userSessions.delete(userId);
      this.bot.sendMessage(msg.chat.id, '❌ Transazione annullata.');
    } else {
      this.bot.sendMessage(msg.chat.id, 'ℹ️ Nessuna transazione in corso da annullare.');
    }
  }
}
